From dbb9895e9fc9f524433dd247adda81fff97f8558 Mon Sep 17 00:00:00 2001
From: Marius Vlad <marius.vlad@collabora.com>
Date: Tue, 30 Mar 2021 11:32:05 +0300
Applied-Upstream: commit: e62ccf179afa98bb29cc35d8c83aea1c37f41409
Subject: [PATCH 2/5] kiosk-shell: Give keyboard focus when mapping the surface

Doing it when the surface is being added would cause clients that
wait for frame callbacks to wait indefinitely as the surface being
activated is not yet, committed.

Fixes: #473
(cherry-picked e62ccf179afa98bb29cc35d8c83aea1c37f41409)
---
 kiosk-shell/kiosk-shell.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/kiosk-shell/kiosk-shell.c b/kiosk-shell/kiosk-shell.c
index a5713a33..ec56025b 100644
--- a/kiosk-shell/kiosk-shell.c
+++ b/kiosk-shell/kiosk-shell.c
@@ -638,16 +638,12 @@ desktop_surface_added(struct weston_desktop_surface *desktop_surface,
 {
 	struct kiosk_shell *shell = data;
 	struct kiosk_shell_surface *shsurf;
-	struct weston_seat *seat;
 
 	shsurf = kiosk_shell_surface_create(shell, desktop_surface);
 	if (!shsurf)
 		return;
 
 	kiosk_shell_surface_set_fullscreen(shsurf, NULL);
-
-	wl_list_for_each(seat, &shell->compositor->seat_list, link)
-		weston_view_activate(shsurf->view, seat, 0);
 }
 
 /* Return the view that should gain focus after the specified shsurf is
@@ -804,6 +800,8 @@ desktop_surface_committed(struct weston_desktop_surface *desktop_surface,
 	}
 
 	if (!weston_surface_is_mapped(surface)) {
+		struct weston_seat *seat;
+
 		struct weston_layer_entry *layer_list =
 			kiosk_shell_verify_apps(shsurf);
 
@@ -811,6 +809,9 @@ desktop_surface_committed(struct weston_desktop_surface *desktop_surface,
 					  &shsurf->view->layer_link);
 		shsurf->view->is_mapped = true;
 		surface->is_mapped = true;
+
+		wl_list_for_each(seat, &shsurf->shell->compositor->seat_list, link)
+			weston_view_activate(shsurf->view, seat, 0);
 	}
 
 	if (!is_fullscreen && (sx != 0 || sy != 0)) {
-- 
2.29.2

